File: Tools/golden/marker_focus.py
#!/usr/bin/env python3
"""162 — **원문이 드러난 줄과 기억이 갈리지 않는가**를 여기서 먼저 따진다.

`MarkerFocus` 는 순수 함수라 CI 의 `swift test` 가 심판이 된다. 다만 **리눅스 세션에서는
Swift 를 못 돌린다** (`CLAUDE.md` §2). 그래서 같은 규칙을 파이썬으로 한 번 더 적고
**경우의 수를 다 돌려** 규칙 자체가 옳은지 여기서 본다.

화면을 **집합**으로 흉내 낸다 — 드러난 줄이 둘 이상일 수 있어야 162 를 표현할 수 있다.
옛 규칙을 같이 넣어 둔 까닭은 **이 시험이 무는지 보이기 위해서**다. 무는 것을 못 본
시험은 통과해도 아무 말을 안 한 것이다.

    python3 Tools/golden/marker_focus.py
"""
import itertools
import sys

# 문단 둘이면 충분하다 — 셋째는 둘째와 같은 말을 한다.
A, B = "A", "B"
CURSORS = [None, A, B]

# 한 걸음에 일어날 수 있는 일.
#   ("plan", 커서, 칠할 수 있나)  — 선택이 바뀌었다 · 초점이 오갔다 · 조합이 끝났다
#   ("whole", 커서)               — 글을 통째로 다시 칠했다 (노트를 열었다)
#   ("nudge",)                    — 다음에는 같은 자리라도 반드시 다시 칠한다
STEPS = (
    [("plan", c, p) for c in CURSORS for p in (True, False)]
    + [("whole", c) for c in CURSORS]
    + [("nudge",)]
)


def plan_new(state, cursor, can_paint):
    dressed, owes = state
    if not can_paint:
        return (None, None), (dressed, owes or dressed != cursor)
    if dressed == cursor and not owes:
        return (None, None), (dressed, False)
    hide = None if dressed == cursor else dressed
    return (hide, cursor), (cursor, False)


def plan_old(state, cursor, can_paint):
    """빌드 52 까지의 규칙 — 칠하지 못했는데 **기억을 지운다.**"""
    dressed, owes = state
    if cursor is None:                      # textViewDidEndEditing
        if can_paint:
            return (dressed, None), (None, False)
        return (None, None), (None, False)  # ← 못 지웠는데 기억만 지운다 (162)
    if not can_paint:
        return (None, None), (dressed, owes)
    if dressed == cursor:
        return (None, None), (dressed, owes)
    return (dressed, cursor), (cursor, False)


def run(rule, steps):
    """규칙대로 걸어 보며 **화면**과 **기억**을 따로 들고 간다."""
    state = (None, False)
    screen = set()
    for step in steps:
        if step[0] == "plan":
            (hide, show), state = rule(state, step[1], step[2])
            if hide is not None:
                screen.discard(hide)
            if show is not None:
                screen.add(show)
        elif step[0] == "whole":
            screen = {step[1]} if step[1] is not None else set()
            state = (step[1], False)
        else:
            state = (state[0], True)

        remembered = {state[0]} if state[0] is not None else set()
        if len(screen) > 1:
            return step, "원문이 드러난 줄이 둘 이상이다", screen, state
        if screen != remembered:
            return step, "화면과 기억이 갈렸다", screen, state
        if step[0] == "plan" and step[2] and state[0] != step[1]:
            return step, "칠할 수 있었는데 커서를 안 따라갔다", screen, state
    return None


def check(name, rule, depth):
    for steps in itertools.product(STEPS, repeat=depth):
        bad = run(rule, steps)
        if bad:
            return name, steps, bad
    return None


def main():
    depth = 4
    total = len(STEPS) ** depth
    print(f"한 걸음에 {len(STEPS)} 가지 · {depth} 걸음 = {total:,} 갈래")

    old = check("옛 규칙", plan_old, depth)
    if old is None:
        print("옛 규칙이 안 걸렸다 — **시험이 안 무는 것이다.** 규칙을 다시 본다.")
        return 1
    _, steps, (step, why, screen, state) = old
    print(f"\n옛 규칙 ✗ {why}")
    print(f"  걸음: {' → '.join(str(s) for s in steps)}")
    print(f"  걸린 자리: {step} · 화면 {sorted(screen)} · 기억 {state}")

    new = check("새 규칙", plan_new, depth)
    if new is not None:
        _, steps, (step, why, screen, state) = new
        print(f"\n새 규칙 ✗ {why}")
        print(f"  걸음: {' → '.join(str(s) for s in steps)}")
        print(f"  걸린 자리: {step} · 화면 {sorted(screen)} · 기억 {state}")
        return 1
    print(f"\n새 규칙 ✓ {total:,} 갈래를 다 걸어도 화면과 기억이 안 갈린다")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: Packages/Core/Sources/Core/Editor/MarkerFocus.swift
import Foundation

/// **지금 어느 줄이 원문을 드러내고 있나** (162, 2026-09-24 사용자 · 아이패드 스크린샷).
///
/// 편집기는 **커서가 있는 문단 하나**만 마크다운 원문으로 보여 준다 (L2). 나머지는
/// 읽는 모습이다. 그런데 한 화면에 **세 줄**이 드러나 있었다 — `# 주간 계획` 의 `#`,
/// 첫 문단의 `**`, `참고자료` 줄의 `[…](…)`.
///
/// **까닭은 기억과 화면이 갈린 것이었다.** 예전에는 `UITextView` 대리자 셋이 각자
/// *어느 문단을 지우고 어느 문단을 드러낼까* 를 판단했고, 드러난 줄을 적어 두는 자리는
/// 하나였는데 **그 자리를 지우면서 화면은 안 지우는 길**이 있었다:
///
/// - 한글은 **늘 조합 중**이다. 조합 도중에 초점이 떠나면 숨기는 일이 건너뛰어진다.
/// - 그런데 바로 다음 줄에서 *드러난 줄이 없다* 고 적어 버렸다.
/// - 그 뒤로는 지울 근거가 사라져 그 줄이 **영영 드러난 채** 남았다. 드러난 줄이 쌓였다.
///
/// `CLAUDE.md` §1 — **같은 것을 재는 곳이 둘이면 언젠가 갈린다.** 여기서는 *어느 줄이
/// 드러났나* 를 화면과 기억이 따로 들고 있었다.
///
/// 그래서 **판단을 여기 하나로 모았다.** 대리자들은 *지금 커서가 어느 문단에 있나* 와
/// *지금 칠해도 되나* 만 말하고, 무엇을 지우고 무엇을 드러낼지는 이 함수가 정한다.
/// 순수 함수라 **리눅스 `swift test` 가 심판**이 된다 — 예전에는 사람 말고 심판이 없었다.
///
/// 오프셋은 **UTF-16** (`NSTextStorage` 단위 — `Formatting` 과 같다).
public enum MarkerFocus {

    /// 문단 하나가 차지하는 자리.
    public struct Span: Equatable, Sendable {
        public let start: Int
        public let length: Int

        public init(start: Int, length: Int) {
            self.start = start
            self.length = length
        }
    }

    /// **지금 화면이 어떤 상태인가.** 편집기가 이것 하나만 들고 있는다.
    public struct State: Equatable, Sendable {
        /// 원문이 드러나 있는 문단. `nil` 이면 드러난 줄이 없다.
        ///
        /// **이 값은 화면을 따라간다.** 실제로 숨긴 뒤에만 `nil` 이 되고, 실제로 드러낸
        /// 뒤에만 값이 들어간다. 숨기지 못했으면 **그대로 들고 있는다** — 그것이 162 다.
        public let dressed: Span?

        /// **갚을 것이 있다.** 칠하지 못하고 지나간 자리가 있었다는 표시.
        /// 다음에 칠할 수 있게 되면 같은 문단이라도 다시 칠한다.
        public let owes: Bool

        public init(dressed: Span? = nil, owes: Bool = false) {
            self.dressed = dressed
            self.owes = owes
        }
    }

    /// **무엇을 칠할 것인가.** 편집기는 이대로만 하면 된다.
    public struct Plan: Equatable, Sendable {
        /// 커서가 없는 모습으로 되돌릴 문단.
        public let hide: Span?
        /// 커서가 있는 모습으로 드러낼 문단.
        public let show: Span?
        /// 칠한 뒤의 상태.
        public let state: State

        public init(hide: Span?, show: Span?, state: State) {
            self.hide = hide
            self.show = show
            self.state = state
        }

        /// 칠할 것이 없다.
        public var isEmpty: Bool { hide == nil && show == nil }
    }

    /// **커서가 여기 있고, 지금 칠할 수 있나.** 그 둘만 주면 나머지는 여기서 정한다.
    ///
    /// - `cursor`: 커서가 든 문단. **초점이 없으면 `nil`** — 드러난 줄이 없어야 한다는 뜻이다.
    /// - `canPaint`: 지금 속성을 건드려도 되나. 한글 조합 중이면 **안 된다** — 건드리면
    ///   조합이 끊긴다 (S10). 그때는 아무것도 칠하지 않고 **갚을 것만 적어 둔다.**
    public static func plan(from state: State, cursor: Span?, canPaint: Bool) -> Plan {
        // 칠할 수 없다 — **기억을 지우지 않는다.** 화면에 드러난 줄은 그대로 있다.
        // 예전에는 여기서 *드러난 줄이 없다* 고 적어 버려 지울 근거를 잃었다 (162).
        guard canPaint else {
            return Plan(hide: nil, show: nil,
                        state: State(dressed: state.dressed,
                                     owes: state.owes || state.dressed != cursor))
        }
        // 이미 맞고 갚을 것도 없으면 가만히 둔다 — 움직일 때마다 칠하면 화면이 떤다 (135).
        guard state.dressed != cursor || state.owes else {
            return Plan(hide: nil, show: nil, state: State(dressed: state.dressed, owes: false))
        }
        let hide = state.dressed == cursor ? nil : state.dressed
        return Plan(hide: hide, show: cursor, state: State(dressed: cursor, owes: false))
    }

    /// **글을 통째로 다시 칠했다** (노트를 열거나 파일이 바뀌어 다시 읽었을 때).
    ///
    /// 그때는 커서가 든 문단이 **이미 드러난 채로 칠해진다.** 그 사실을 적어 두지 않으면
    /// 화면과 기억이 또 갈린다 — 드러난 줄을 아무도 못 지우게 된다.
    public static func afterWholeRepaint(cursor: Span?) -> State {
        State(dressed: cursor, owes: false)
    }

    /// **다음에는 같은 자리라도 반드시 다시 칠한다.**
    ///
    /// 초점이 돌아올 때 쓴다. 대리자가 불리는 차례 때문에 커서 자리가 아직 안 정해진
    /// 순간이 있어(빌드 31 · 100), 그 뒤에 오는 칠하기가 **건너뛰지 않게** 해야 한다.
    /// 예전에는 기억을 지워서 그 일을 했는데, 지우면 드러난 줄을 잃는다 (162).
    public static func nudged(_ state: State) -> State {
        State(dressed: state.dressed, owes: true)
    }
}
